File: android/ares/src/main/java/com/xt/ares/ARESJSEnvContext.kt
package com.xt.ares

import android.R.attr.max
import android.graphics.PathMeasure
import android.graphics.Region
import com.xt.ares.commands.*
import org.mozilla.javascript.ScriptableObject
import android.R.attr.path
import android.graphics.RectF
import org.json.JSONObject

class ARESJSEnvContext(val view: ARESView): ScriptableObject() {

    init {
        this.defineProperty(
                "canvas",
                null,
                ARESJSEnvContext::class.java.getDeclaredMethod("getCanvas"),
                null,
                0)
        this.defineProperty(
                "globalAlpha",
                null,
                ARESJSEnvContext::class.java.getDeclaredMethod("getGlobalAlpha"),
                ARESJSEnvContext::class.java.getDeclaredMethod("setGlobalAlpha", Double::class.java),
                0)
        this.defineProperty(
                "globalCompositeOperation",
                null,
                ARESJSEnvContext::class.java.getDeclaredMethod("getGlobalCompositeOperation"),
                ARESJSEnvContext::class.java.getDeclaredMethod("setGlobalCompositeOperation", String::class.java),
                0)
        this.defineProperty(
                "fillStyle",
                null,
                ARESJSEnvContext::class.java.getDeclaredMethod("getFillStyle"),
                ARESJSEnvContext::class.java.getDeclaredMethod("setFillStyle", String::class.java),
                0)
        this.defineProperty(
                "strokeStyle",
                null,
                ARESJSEnvContext::class.java.getDeclaredMethod("getStrokeStyle"),
                ARESJSEnvContext::class.java.getDeclaredMethod("setStrokeStyle", String::class.java),
                0)
        this.defineProperty(
                "lineWidth",
                null,
                ARESJSEnvContext::class.java.getDeclaredMethod("getLineWidth"),
                ARESJSEnvContext::class.java.getDeclaredMethod("setLineWidth", Double::class.java),
                0)
        this.defineProperty(
                "lineCap",
                null,
                ARESJSEnvContext::class.java.getDeclaredMethod("getLineCap"),
                ARESJSEnvContext::class.java.getDeclaredMethod("setLineCap", String::class.java),
                0)
        this.defineProperty(
                "lineJoin",
                null,
                ARESJSEnvContext::class.java.getDeclaredMethod("getLineJoin"),
                ARESJSEnvContext::class.java.getDeclaredMethod("setLineJoin", String::class.java),
                0)
        this.defineProperty(
                "miterLimit",
                null,
                ARESJSEnvContext::class.java.getDeclaredMethod("getMiterLimit"),
                ARESJSEnvContext::class.java.getDeclaredMethod("setMiterLimit", Double::class.java),
                0)
        this.defineProperty(
                "shadowOffsetX",
                null,
                ARESJSEnvContext::class.java.getDeclaredMethod("getShadowOffsetX"),
                ARESJSEnvContext::class.java.getDeclaredMethod("setShadowOffsetX", Double::class.java),
                0)
        this.defineProperty(
                "shadowOffsetY",
                null,
                ARESJSEnvContext::class.java.getDeclaredMethod("getShadowOffsetY"),
                ARESJSEnvContext::class.java.getDeclaredMethod("setShadowOffsetY", Double::class.java),
                0)
        this.defineProperty(
                "shadowBlur",
                null,
                ARESJSEnvContext::class.java.getDeclaredMethod("getShadowBlur"),
                ARESJSEnvContext::class.java.getDeclaredMethod("setShadowBlur", Double::class.java),
                0)
        this.defineProperty(
                "shadowColor",
                null,
                ARESJSEnvContext::class.java.getDeclaredMethod("getShadowColor"),
                ARESJSEnvContext::class.java.getDeclaredMethod("setShadowColor", String::class.java),
                0)
        this.defineProperty(
                "font",
                null,
                ARESJSEnvContext::class.java.getDeclaredMethod("getFont"),
                ARESJSEnvContext::class.java.getDeclaredMethod("setFont", String::class.java),
                0)
        this.defineProperty(
                "textAlign",
                null,
                ARESJSEnvContext::class.java.getDeclaredMethod("getTextAlign"),
                ARESJSEnvContext::class.java.getDeclaredMethod("setTextAlign", String::class.java),
                0)
        this.defineProperty(
                "textBaseline",
                null,
                ARESJSEnvContext::class.java.getDeclaredMethod("getTextBaseline"),
                ARESJSEnvContext::class.java.getDeclaredMethod("setTextBaseline", String::class.java),
                0)
        this.defineFunctionProperties(arrayOf(
                "save",
                "restore",
                "scale",
                "rotate",
                "translate",
                "transform",
                "setTransform",
                "fillRect",
                "strokeRect",
                "clearRect",
                "beginPath",
                "closePath",
                "stroke",
                "fill",
                "lineTo",
                "moveTo",
                "rect",
                "quadraticCurveTo",
                "bezierCurveTo",
                "arc",
                "arcTo",
                "clip",
                "isPointInPath",
                "fillText",
                "strokeText",
                "measureText",
                "drawImage",
                "update"
                ), ARESJSEnvContext::class.java, 0)
    }

    override fun getClassName(): String {
        return "Context"
    }

    var canvas: ARESJSCanvasObject? = null
        get() {
            return ARESJSCanvasObject(this.view, this)
        }

    fun save() {
        this.view.addCommand(ARESSaveCommand())
    }

    fun restore() {
        this.view.addCommand(ARESRestoreCommand())
    }

    fun scale(x: Double, y: Double) {
        this.view.addCommand(ARESScaleCommand(x, y))
    }

    fun rotate(angle: Double) {
        this.view.addCommand(ARESRotateCommand(angle))
    }

    fun translate(x: Double, y: Double) {
        this.view.addCommand(ARESTranslateCommand(x, y))
    }

    fun transform(a: Double, b: Double, c: Double, d: Double, tx: Double, ty: Double) {
        this.view.addCommand(ARESTransformCommand(a, b, c, d, tx, ty))
    }

    fun setTransform(a: Double, b: Double, c: Double, d: Double, tx: Double, ty: Double) {
        this.view.addCommand(ARESSetTransformCommand(a, b, c, d, tx, ty))
    }

    var globalAlpha: Double = 1.0
        set(value) {
            field = value
            this.view.addCommand(ARESGlobalAlphaCommand(value))
        }

    var globalCompositeOperation: String = ""
        set(value) {
            field = value
            this.view.addCommand(ARESGlobalCompositeOperationCommand(value))
        }

    var fillStyle: String = ""
        set(value) {
            field = value
            this.view.addCommand(ARESFillStyleCommand(value))
        }

    var strokeStyle: String = ""
        set(value) {
            field = value
            this.view.addCommand(ARESStrokeStyleCommand(value))
        }

    var lineWidth: Double = 1.0
        set(value) {
            field = value
            this.view.addCommand(ARESLineWidthCommand(value))
        }

    var lineCap: String = ""
        set(value) {
            field = value
            this.view.addCommand(ARESLineCapCommand(value))
        }

    var lineJoin: String = ""
        set(value) {
            field = value
            this.view.addCommand(ARESLineJoinCommand(value))
        }

    var miterLimit: Double = 0.0
        set(value) {
            field = value
            this.view.addCommand(ARESMiterLimitCommand(value))
        }

    var shadowOffsetX: Double = 0.0
        set(value) {
            field = value
            this.view.addCommand(ARESShadowCommand(value, this.shadowOffsetY, this.shadowBlur, this.shadowColor))
        }

    var shadowOffsetY: Double = 0.0
        set(value) {
            field = value
            this.view.addCommand(ARESShadowCommand(this.shadowOffsetX, value, this.shadowBlur, this.shadowColor))
        }

    var shadowBlur: Double = 0.0
        set(value) {
            field = value
            this.view.addCommand(ARESShadowCommand(this.shadowOffsetX, this.shadowOffsetY, value, this.shadowColor))
        }

    var shadowColor: String = ""
        set(value) {
            field = value
            this.view.addCommand(ARESShadowCommand(this.shadowOffsetX, this.shadowOffsetY, this.shadowBlur, value))
        }

    var font: String = "10px"
        set(value) {
            field = value
            this.view.addCommand(ARESFontCommand(value))
        }

    var textAlign: String = "left"
        set(value) {
            field = value
            this.view.addCommand(ARESTextAlignCommand(value))
        }

    var textBaseline: String = ""
        set(value) {
            field = value
            this.view.addCommand(ARESTextBaselineCommand(value))
        }

    fun fillRect(x: Double, y: Double, w: Double, h: Double) {
        this.view.addCommand(ARESFillRectCommand(x, y, w, h))
    }

    fun strokeRect(x: Double, y: Double, w: Double, h: Double) {
        this.view.addCommand(ARESStrokeRectCommand(x, y, w, h))
    }

    fun clearRect(x: Double, y: Double, w: Double, h: Double) {
        this.view.addCommand(ARESClearRectCommand(x, y, w, h))
    }

    fun beginPath() {
        this.view.addCommand(ARESBeginPathCommand())
    }

    fun closePath() {
        this.view.addCommand(ARESClosePathCommand())
    }

    fun stroke() {
        this.view.addCommand(ARESStrokeCommand())
    }

    fun fill() {
        this.view.addCommand(ARESFillCommand())
    }

    fun lineTo(x: Double, y: Double) {
        this.view.addCommand(ARESLineToCommand(x, y))
    }

    fun moveTo(x: Double, y: Double) {
        this.view.addCommand(ARESMoveToCommand(x, y))
    }

    fun rect(x: Double, y: Double, w: Double, h: Double) {
        this.view.addCommand(ARESRectCommand(x, y, w, h))
    }

    fun quadraticCurveTo(cpx: Double, cpy: Double, x: Double, y: Double) {
        this.view.addCommand(ARESQuadraticCurveToCommand(cpx, cpy, x, y))
    }

    fun bezierCurveTo(cp1x: Double, cp1y: Double, cp2x: Double, cp2y: Double, x: Double, y: Double) {
        this.view.addCommand(ARESBezierCurveToCommand(cp1x, cp1y, cp2x, cp2y, x, y))
    }

    fun arc(x: Double, y: Double, r: Double, start: Double, end: Double, anticlockwise: Boolean) {
        this.view.addCommand(ARESArcCommand(x, y, r, start, end, anticlockwise))
    }

    fun arcTo(x1: Double, y1: Double, x2: Double, y2: Double, radius: Double) {
        this.view.addCommand(ARESArcToCommand(x1, y1, x2, y2, radius))
    }

    fun clip() {
        this.view.addCommand(ARESClipCommand())
    }

    fun isPointInPath(x: Double, y: Double): Boolean {
        val rectF = RectF()
        ARESBeginPathCommand.sharedPath.computeBounds(rectF, true)
        val region = Region()
        region.setPath(ARESBeginPathCommand.sharedPath, Region(rectF.left.toInt(), rectF.top.toInt(), rectF.right.toInt(), rectF.bottom.toInt()))
        return region.contains(x.toInt(), y.toInt())
    }

    fun fillText(text: String, x: Double, y: Double, maxWidth: Double) {
        this.view.addCommand(ARESFillTextCommand(text, x, y, maxWidth))
    }

    fun strokeText(text: String, x: Double, y: Double, maxWidth: Double) {
        this.view.addCommand(ARESStrokeTextCommand(text, x, y, maxWidth))
    }

    fun measureText(text: String): TextMeasureResult {
        aresStateFont?.let {
            view.currentTextPaint.textSize = it.textSize
            view.currentTextPaint.typeface = it.typeface
            return TextMeasureResult(view.currentTextPaint.measureText(text).toDouble())
        }
        return TextMeasureResult(0.0)
    }

    fun drawImage(image: ARESImage, sxOrDx: Double, syOrDy: Double, swOrDw: Double, shOrDh: Double, dx: Double, dy: Double, dw: Double, dh: Double) {
        if (!dx.isNaN()) {
            this.view.addCommand(ARESDrawImageCommand(
                    image,
                    dx,
                    dy,
                    dw,
                    dh,
                    sxOrDx,
                    syOrDy,
                    swOrDw,
                    shOrDh
            ))
        }
        else {
            this.view.addCommand(ARESDrawImageCommand(
                    image,
                    sxOrDx,
                    syOrDy,
                    swOrDw,
                    shOrDh,
                    Double.NaN,
                    Double.NaN,
                    Double.NaN,
                    Double.NaN
            ))
        }
    }

    fun update() {
        this.view.update()
    }

    class TextMeasureResult(val width: Double)

}

class ARESJSCanvasObject(val view: ARESView, val ctx: ARESJSEnvContext): ScriptableObject() {

    init {
        this.defineProperty(
                "width",
                null,
                ARESJSCanvasObject::class.java.getDeclaredMethod("getWidth"),
                null,
                0)
        this.defineProperty(
                "height",
                null,
                ARESJSCanvasObject::class.java.getDeclaredMethod("getHeight"),
                null,
                0)
        this.defineFunctionProperties(arrayOf(
                "getContext"
        ), ARESJSCanvasObject::class.java, 0)
    }

    override fun getClassName(): String {
        return "Canvas"
    }

    var width: Double = 0.0
        get() {
            return (this.view.width / this.view.resources.displayMetrics.density).toDouble()
        }

    var height: Double = 0.0
        get() {
            return (this.view.height / this.view.resources.displayMetrics.density).toDouble()
        }

    fun getContext(): ARESJSEnvContext {
        return this.ctx
    }

}

File: ios/sources/ARESJSEnvContext.m
//
//  ARESJSEnvContext.m
//  ares
//
//  Created by PonyCui on 2018/11/28.
//  Copyright © 2018年 XT Studio. All rights reserved.
//

#import "ARESJSEnvContext.h"
#import "ARESJSBridge.h"
#import "ARESView.h"
#import "ARESSaveCommand.h"
#import "ARESRestoreCommand.h"
#import "ARESScaleCommand.h"
#import "ARESRotateCommand.h"
#import "ARESTranslateCommand.h"
#import "ARESTransformCommand.h"
#import "ARESSetTransformCommand.h"
#import "ARESGlobalAlphaCommand.h"
#import "ARESGlobalCompositeOperationCommand.h"
#import "ARESFillStyleCommand.h"
#import "ARESStrokeStyleCommand.h"
#import "ARESLineWidthCommand.h"
#import "ARESLineCapCommand.h"
#import "ARESLineJoinCommand.h"
#import "ARESMiterLimitCommand.h"
#import "ARESShadowCommand.h"
#import "ARESFillRectCommand.h"
#import "ARESStrokeRectCommand.h"
#import "ARESClearRectCommand.h"
#import "ARESBeginPathCommand.h"
#import "ARESClosePathCommand.h"
#import "ARESStrokeCommand.h"
#import "ARESFillCommand.h"
#import "ARESLineToCommand.h"
#import "ARESMoveToCommand.h"
#import "ARESRectCommand.h"
#import "ARESQuadraticCurveToCommand.h"
#import "ARESBezierCurveToCommand.h"
#import "ARESArcCommand.h"
#import "ARESArcToCommand.h"
#import "ARESClipCommand.h"
#import "ARESFontCommand.h"
#import "ARESTextAlignCommand.h"
#import "ARESTextBaselineCommand.h"
#import "ARESFillTextCommand.h"
#import "ARESStrokeTextCommand.h"
#import "ARESMeasureTextCommand.h"
#import "ARESDrawImageCommand.h"
#import "ARESJSImage.h"

@implementation ARESJSEnvContext

- (instancetype)init
{
    self = [super init];
    if (self) {
        _lineWidth = 1.0;
        _font = @"10px sans-serif";
    }
    return self;
}

- (NSDictionary *)canvas {
    ARESView *view = [ARESJSBridge viewWithContext:[JSContext currentContext]];
    return @{
             @"width": @(view.bounds.size.width),
             @"height": @(view.bounds.size.height),
             @"getContext": ^ {
                 return self;
             },
             };
}

- (void)save {
    [self addCommandToView:[[ARESSaveCommand alloc] init]];
}

- (void)restore {
    [self addCommandToView:[[ARESRestoreCommand alloc] init]];
}

- (void)scale:(float)x y:(float)y {
    [self addCommandToView:[[ARESScaleCommand alloc] initWithSx:x sy:y]];
}

- (void)rotate:(float)angle {
    [self addCommandToView:[[ARESRotateCommand alloc] initWithAngle:angle]];
}

- (void)translate:(float)x y:(float)y {
    [self addCommandToView:[[ARESTranslateCommand alloc] initWithTx:x ty:y]];
}

- (void)transform:(float)a b:(float)b c:(float)c d:(float)d tx:(float)tx ty:(float)ty {
    [self addCommandToView:[[ARESTransformCommand alloc] initWithA:a b:b c:c d:d tx:tx ty:ty]];
}

- (void)setTransform:(float)a b:(float)b c:(float)c d:(float)d tx:(float)tx ty:(float)ty {
    [self addCommandToView:[[ARESSetTransformCommand alloc] initWithA:a b:b c:c d:d tx:tx ty:ty]];
}

- (void)setGlobalAlpha:(float)globalAlpha {
    _globalAlpha = globalAlpha;
    [self addCommandToView:[[ARESGlobalAlphaCommand alloc] initWithValue:globalAlpha]];
}

- (void)setGlobalCompositeOperation:(NSString *)globalCompositeOperation {
    _globalCompositeOperation = globalCompositeOperation;
    [self addCommandToView:[[ARESGlobalCompositeOperationCommand alloc] initWithValue:globalCompositeOperation]];
}

- (void)setFillStyle:(NSString *)fillStyle {
    _fillStyle = fillStyle;
    [self addCommandToView:[[ARESFillStyleCommand alloc] initWithValue:fillStyle]];
}

- (void)setStrokeStyle:(NSString *)strokeStyle {
    _strokeStyle = strokeStyle;
    [self addCommandToView:[[ARESStrokeStyleCommand alloc] initWithValue:strokeStyle]];
}

- (void)setLineWidth:(float)lineWidth {
    _lineWidth = lineWidth;
    [self addCommandToView:[[ARESLineWidthCommand alloc] initWithLineWidth:lineWidth]];
}

- (void)setLineCap:(NSString *)lineCap {
    _lineCap = lineCap;
    [self addCommandToView:[[ARESLineCapCommand alloc] initWithValue:lineCap]];
}

- (void)setLineJoin:(NSString *)lineJoin {
    _lineJoin = lineJoin;
    [self addCommandToView:[[ARESLineJoinCommand alloc] initWithValue:lineJoin]];
}

- (void)setMiterLimit:(float)miterLimit {
    _miterLimit = miterLimit;
    [self addCommandToView:[[ARESMiterLimitCommand alloc] initWithValue:miterLimit]];
}

- (void)setShadowOffsetX:(float)shadowOffsetX {
    _shadowOffsetX = shadowOffsetX;
    [self addCommandToView:[[ARESShadowCommand alloc] initWithX:shadowOffsetX
                                                              y:self.shadowOffsetY
                                                           blur:self.shadowBlur
                                                    colorString:self.shadowColor]];
}

- (void)setShadowOffsetY:(float)shadowOffsetY {
    _shadowOffsetY = shadowOffsetY;
    [self addCommandToView:[[ARESShadowCommand alloc] initWithX:self.shadowOffsetX
                                                              y:shadowOffsetY
                                                           blur:self.shadowBlur
                                                    colorString:self.shadowColor]];
}

- (void)setShadowBlur:(float)shadowBlur {
    _shadowBlur = shadowBlur;
    [self addCommandToView:[[ARESShadowCommand alloc] initWithX:self.shadowOffsetX
                                                              y:self.shadowOffsetY
                                                           blur:shadowBlur
                                                    colorString:self.shadowColor]];
}

- (void)setShadowColor:(NSString *)shadowColor {
    _shadowColor = shadowColor;
    [self addCommandToView:[[ARESShadowCommand alloc] initWithX:self.shadowOffsetY
                                                              y:self.shadowOffsetY
                                                           blur:self.shadowBlur
                                                    colorString:shadowColor]];
}

- (void)fillRect:(float)x y:(float)y w:(float)w h:(float)h {
    [self addCommandToView:[[ARESFillRectCommand alloc]
                            initWithX:x
                            y:y
                            w:w
                            h:h]];
}

- (void)strokeRect:(float)x y:(float)y w:(float)w h:(float)h {
    [self addCommandToView:[[ARESStrokeRectCommand alloc]
                            initWithX:x
                            y:y
                            w:w
                            h:h]];
}

- (void)clearRect:(float)x y:(float)y w:(float)w h:(float)h {
    [self addCommandToView:[[ARESClearRectCommand alloc]
                            initWithX:x
                            y:y
                            w:w
                            h:h]];
}

- (void)beginPath {
    [self addCommandToView:[[ARESBeginPathCommand alloc] init]];
}

- (void)closePath {
    [self addCommandToView:[[ARESClosePathCommand alloc] init]];
}

- (void)stroke {
    [self addCommandToView:[[ARESStrokeCommand alloc] init]];
}

- (void)fill {
    [self addCommandToView:[[ARESFillCommand alloc] init]];
}

- (void)lineTo:(float)x y:(float)y {
    [self addCommandToView:[[ARESLineToCommand alloc] initWithX:x y:y]];
}

- (void)moveTo:(float)x y:(float)y {
    [self addCommandToView:[[ARESMoveToCommand alloc] initWithX:x y:y]];
}

- (void)rect:(float)x y:(float)y w:(float)w h:(float)h {
    [self addCommandToView:[[ARESRectCommand alloc] initWithX:x y:y w:w h:h]];
}

- (void)quadraticCurveTo:(float)cpx cpy:(float)cpy x:(float)x y:(float)y {
    [self addCommandToView:[[ARESQuadraticCurveToCommand alloc] initWithCpx:cpx cpy:cpy x:x y:y]];
}

- (void)bezierCurveTo:(float)cp1x cp1y:(float)cp1y cp2x:(float)cp2x cp2y:(float)cp2y x:(float)x y:(float)y {
    [self addCommandToView:[[ARESBezierCurveToCommand alloc] initWithCp1x:cp1x
                                                                     cp1y:cp1y
                                                                     cp2x:cp2x
                                                                     cp2y:cp2y
                                                                        x:x
                                                                        y:y]];
}

- (void)arc:(float)x y:(float)y r:(float)r start:(float)start end:(float)end anticlockwise:(BOOL)anticlockwise {
    [self addCommandToView:[[ARESArcCommand alloc] initWithX:x
                                                           y:y
                                                           r:r
                                                       start:start
                                                         end:end
                                               anticlockwise:anticlockwise]];
}

- (void)arcTo:(float)x1 y1:(float)y1 x2:(float)x2 y2:(float)y2 radius:(float)radius {
    [self addCommandToView:[[ARESArcToCommand alloc] initWithX1:x1 y1:y1 x2:x2 y2:y2 radius:radius]];
}

- (void)clip {
    [self addCommandToView:[[ARESClipCommand alloc] init]];
}

- (BOOL)isPointInPath:(float)x y:(float)y {
    return [[ARESBeginPathCommand currentPath] containsPoint:CGPointMake(x, y)];
}

- (void)setFont:(NSString *)font {
    _font = font;
    [self addCommandToView:[[ARESFontCommand alloc] initWithValue:font]];
}

- (void)setTextAlign:(NSString *)textAlign {
    _textAlign = textAlign;
    [self addCommandToView:[[ARESTextAlignCommand alloc] initWithValue:textAlign]];
}

- (void)setTextBaseline:(NSString *)textBaseline {
    _textBaseline = textBaseline;
    [self addCommandToView:[[ARESTextBaselineCommand alloc] initWithValue:textBaseline]];
}

- (void)fillText:(NSString *)text x:(float)x y:(float)y maxWidth:(float)maxWidth {
    [self addCommandToView:[[ARESFillTextCommand alloc] initWithText:text
                                                                   x:x
                                                                   y:y
                                                            maxWidth:maxWidth]];
}

- (void)strokeText:(NSString *)text x:(float)x y:(float)y maxWidth:(float)maxWidth {
    [self addCommandToView:[[ARESStrokeTextCommand alloc] initWithText:text
                                                                     x:x
                                                                     y:y
                                                              maxWidth:maxWidth]];
}

- (NSDictionary *)measureText:(NSString *)text {
    return [ARESMeasureTextCommand measureText:text];
}

- (void)drawImage:(ARESJSImage *)image
           sxOrDx:(float)sxOrDx
           syOrDy:(float)syOrDy
           swOrDw:(float)swOrDw
           shOrDh:(float)shOrDh
               dx:(float)dx
               dy:(float)dy
               dw:(float)dw
               dh:(float)dh {
    if (!isnan(dx)) {
        [self addCommandToView:[[ARESDrawImageCommand alloc] initWithImage:image.image
                                                                        dx:dx
                                                                        dy:dy
                                                                        dw:dw
                                                                        dh:dh
                                                                        sx:sxOrDx
                                                                        sy:syOrDy
                                                                        sw:swOrDw
                                                                        sh:shOrDh]];
    }
    else {
        [self addCommandToView:[[ARESDrawImageCommand alloc] initWithImage:image.image
                                                                        dx:sxOrDx
                                                                        dy:syOrDy
                                                                        dw:swOrDw
                                                                        dh:shOrDh
                                                                        sx:NAN
                                                                        sy:NAN
                                                                        sw:NAN
                                                                        sh:NAN]];
    }
}

- (void)addCommandToView:(ARESCommand *)command {
    if (command == nil) {
        return;
    }
    ARESView *view = [ARESJSBridge viewWithContext:[JSContext currentContext]];
    if (view != nil) {
        [view addCommand:command];
    }
}

@end
